src/hardware/gameblaster.cpp

```diff
@@ -16,105 +17,178 @@
  *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
  */
 
-#include "dosbox.h"
-#include "inout.h"
-#include "mixer.h"
-#include "mem.h"
-#include "hardware.h"
+#include "gameblaster.h"
+
 #include "setup.h"
 #include "support.h"
 #include "pic.h"
-#include <cstring>
-#include <math.h>
 
-#include "mame/emu.h"
-#include "mame/saa1099.h"
+void GameBlaster::Open(const int port_choice, const std::string_view card_choice)
+{
+	Close();
+	assert(!is_enabled);
+	assert(!is_open);
 
-// GameBlaster runs at half of ISA's clock speed (14318180 / 2)
-constexpr uint32_t GAMEBLASTER_CLOCK_HZ = 7159090;
+	is_standalone_gameblaster = (card_choice == "gb");
+	const auto card_name = is_standalone_gameblaster ? "GAMEBLASTER" : "CMS";
 
+	// Is the user's port choice valid?
+	if (port_choice < 0x210 || port_choice > 0x260) {
```

Comment:
Would something like 211h or 255h be valid? Shouldn't we just rather whitelist the 210h, 220, 230, etc.?

Reply:
Added allow-lists for GB and CMS modes.